src/core/chuck_vm.cpp

```diff
@@ -771,55 +776,63 @@ t_CKUINT Chuck_VM::process_msg( Chuck_Msg * & msg )
     if( msg->type == CK_MSG_REPLACE )
     {
         Chuck_VM_Shred * out = m_shreduler->lookup( msg->param );
-        if( !out )
+        if( !out && !msg->alwaysAdd )
         {
             EM_print2orange( "(VM) replacing shred: no shred with id %i...", msg->param );
             retval = 0;
             goto done;
         }
 
+        // get a shred one way or another
         Chuck_VM_Shred * shred = msg->shred;
+        // if shred wasn't created on the outside
         if( !shred )
         {
             shred = new Chuck_VM_Shred;
             shred->vm_ref = this;
             shred->initialize( msg->code );
             shred->name = msg->code->name;
             shred->base_ref = shred->mem;
-            shred->add_ref();
+            // NOTE no add_ref() since we spork() later
```

Comment:
indeed, this part of the code in REPLACE was basically "manually" effecting a spork, including the reference count; since we added a spork() in the add-replace/alwaysAdd scenario, the ref count was duplicated. this would leave the shred dangling after it is removed (and won't perform critical cleanup like disconnecting its UGens). For example, removing blit2.ck added in this manner removed the shred from VM but the ugens kept sounding (without the changes in frequency); this has since been refactored to correctly (hopefully) add ref on the new shred

Reply:
Thanks! There's actually a bug with static classes/local ugens mixing that might be related to this dangling shred issue, will dig a littler deeper and post a bug later